graph-builder/src/registry.rs

```diff
@@ -42,55 +42,148 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
-fn trim_protocol(src: &str) -> &str {
-    src.trim_left_matches("https://")
-        .trim_left_matches("http://")
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct Registry {
+    pub(crate) scheme: String,
+    pub(crate) insecure: bool,
+    pub(crate) host: String,
+    pub(crate) port: Option<u16>,
+}
+
+impl Registry {
+    pub fn try_from_str(src: &str) -> Fallible<Self> {
+        if let Some(capture) = regex::Regex::new(
+            // match scheme://h.o.s.t:port
+            r"^(?P<scheme>[a-z]+)(:/{2})(?P<host>([0-9a-zA-Z]+\.)*([0-9a-zA-Z]+)):(?P<port>[0-9]+)$"
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            let scheme = capture["scheme"].to_string();
+            return Ok(Registry {
+                insecure: Registry::insecure_scheme(&scheme)?,
+                scheme,
+                host: capture["host"].to_string(),
+                port: Some(
+                    capture["port"]
+                        .parse()
+                        .expect("could not parse port as a number"),
+                ),
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match scheme://h.o.s.t
+            r"^(?P<scheme>[a-z]+)(:/{2})(?P<host>([0-9a-zA-Z]+\.)*([0-9a-zA-Z]+))$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            let scheme = capture["scheme"].to_string();
+            return Ok(Registry {
+                insecure: Registry::insecure_scheme(&scheme)?,
+                scheme,
+                host: capture["host"].to_string(),
+                ..Default::default()
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match h.o.s.t:port
+            r"^(?P<host>([0-9a-zA-Z\-]+\.)+([0-9a-zA-Z]+)):(?P<port>[0-9]+)$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            return Ok(Registry {
+                host: capture["host"].to_string(),
+                port: Some(
+                    capture["port"]
+                        .parse()
+                        .expect("could not parse port as a number"),
+                ),
+                ..Default::default()
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match h.o.s.t
+            r"^(?P<host>([0-9a-zA-Z\-]+\.)*([0-9a-zA-Z]+))$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
```

Comment:
Personal opinion: I had a brainfart reading this, maybe because there are five lines between the `if let` and its `{`. Naming captures result and having a single line `if let ... {` would make it straightforward to read. (But maybe I just need more coffee :)

Reply:
Since we've agreed on the approach I've refactored the duplicate code into a macro which should further obfuscate yet address this ;-)